File: EclipseKotlin/Kotlin/src/fun/fizzy/patterns/PatternsDriver.kt
package `fun`.fizzy.patterns

/** @desc patterns driver is meant for exploring patterns */
class PatternsDriver: MaObserver {
	/* Action performed by the observer.*/
	override fun update(mamsg : String) {
		println(mamsg)
	}
}

/**
 * Observer.
 */
interface MaObserver {
	/* Action performed by the observer.*/
	fun update(mamsg : String)
}

/**
 * ObserverObject.
 */
class ObservingObject {
	/* List that contains the observers.*/
	private val maListOfObservers: MutableList<MaObserver> = mutableListOf<MaObserver>()

	/* Method used to add an observer to the observers.*/
	fun addObserver(observer: MaObserver) {
		this.maListOfObservers.add(observer)
	}

	/* Method used to remove an observer from the observers.*/
	fun removeObserver(observer: MaObserver){
		this.maListOfObservers.remove(observer)
	}

	/* Method used to notify an observer.*/
	fun notifyObservers(mamsg : String){
		this.maListOfObservers.forEach {
			it -> it.update(mamsg)
		}
	}
}

File: EclipseCpp/Cpp/src/algorithms/patterns/PatternsDriver.cpp
/* @author: pablorosas17
 * @date: 01/25/2025
 * @purpose: implementation of patterns.
 */

#include "Patterns.h"

void PatternsDriver::update(const std::string& mamsg) {
    std::cout << mamsg << std::endl;
}

void ObservingObject::addObserver(const std::shared_ptr<MaObserver>& observer) {
    maListOfObservers.push_back(observer);
}

void ObservingObject::removeObserver(const std::shared_ptr<MaObserver>& observer) {
    maListOfObservers.erase(
        std::remove(maListOfObservers.begin(), maListOfObservers.end(), observer),
        maListOfObservers.end()
    );
}

void ObservingObject::notifyObservers(const std::string& mamsg) {
    for (const auto& observer : maListOfObservers) {
        observer->update(mamsg);
    }
}
